lxd/db/query/dump.go

```diff
@@ -4,96 +4,116 @@ import (
-	schemas := dumpParseSchema(schema)
-
-	// Begin
-	dump := `PRAGMA foreign_keys=OFF;
-BEGIN TRANSACTION;
-`
-	// Schema table
-	tableDump, err := dumpTable(ctx, tx, "schema", dumpSchemaTable)
+// sqlite3's dump feature.
+func Dump(ctx context.Context, tx *sql.Tx, schemaOnly bool) (string, error) {
+	tablesSchemas, tableNames, err := getTablesSchemas(ctx, tx)
 	if err != nil {
-		return "", fmt.Errorf("Failed to dump table schema: %w", err)
+		return "", err
 	}
-	dump += tableDump
 
-	// All other tables
-	tables := make([]string, 0)
-	for table := range schemas {
-		tables = append(tables, table)
-	}
-	sort.Strings(tables)
-	for _, table := range tables {
-		if schemaOnly {
-			// Dump only the schema.
-			dump += schemas[table] + "\n"
-			continue
-		}
+	// Begin dump string.
+	var builder strings.Builder
+	builder.WriteString("PRAGMA foreign_keys=OFF;\n")
+	builder.WriteString("BEGIN TRANSACTION;\n")
 
-		tableDump, err := dumpTable(ctx, tx, table, schemas[table])
-		if err != nil {
-			return "", fmt.Errorf("Failed to dump table %s: %w", table, err)
-		}
+	// For each table, write the schema and optionally write the data.
+	for _, tableName := range tableNames {
+		builder.WriteString(tablesSchemas[tableName] + "\n")
 
-		dump += tableDump
+		if !schemaOnly {
+			tableData, err := getTableData(ctx, tx, tableName)
+			if err != nil {
+				return "", err
+			}
+
+			for _, stmt := range tableData {
+				builder.WriteString(stmt + "\n")
+			}
+		}
 	}
 
-	// Sequences (unless the schemaOnly flag is true)
+	// Sequences (unless the schemaOnly flag is true).
 	if !schemaOnly {
-		tableDump, err = dumpTable(ctx, tx, "sqlite_sequence", "DELETE FROM sqlite_sequence;")
+		builder.WriteString("DELETE FROM sqlite_sequence;\n")
+
+		tableData, err := getTableData(ctx, tx, "sqlite_sequence")
 		if err != nil {
 			return "", fmt.Errorf("Failed to dump table sqlite_sequence: %w", err)
 		}
 
-		dump += tableDump
+		for _, stmt := range tableData {
+			builder.WriteString(stmt + "\n")
+		}
 	}
 
-	// Commit
-	dump += "COMMIT;\n"
+	// Commit.
+	builder.WriteString("COMMIT;\n")
 
-	return dump, nil
+	return builder.String(), nil
 }
 
-// Return a map from table names to their schema definition, taking a full
-// schema SQL text generated with schema.Schema.Dump().
-func dumpParseSchema(schema string) map[string]string {
-	tables := map[string]string{}
-	for _, statement := range strings.Split(schema, ";") {
-		statement = strings.Trim(statement, " \n") + ";"
-		if !strings.HasPrefix(statement, "CREATE TABLE") {
-			continue
+// getTablesSchemas gets all the tables and their schema, as well as a list of table names in their default order from
+// the sqlite_master table.
+func getTablesSchemas(ctx context.Context, tx *sql.Tx) (map[string]string, []string, error) {
+	rows, err := tx.QueryContext(ctx, `SELECT name, sql FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'`)
```

Comment:
Should we use an explicit ORDER BY here to make it clear this should be returned in created order.

Reply:
`ORDER BY rowid;` should cover it.

Reply:
I've sorted by `rowid` now.